Add support for rotation argument handling.

NonSwag opened the pull request:
Resolves #11395 

Introduced a `RotationResolver` to handle rotation arguments and added related methods to `ArgumentTypes`, and argument provider classes. This enables resolving and processing rotation data using `CommandSourceStack`.

The `Vector2f` in `RotationArgument` was just a temp solution because I wasn't sure on whether to add a rotation object that clearly combines a yaw and pitch

Commit:
Add support for rotation argument handling.

Introduced a `RotationResolver` to handle rotation arguments and added related methods to `ArgumentTypes`, and argument provider classes. This enables resolving and processing rotation data using `CommandSourceStack`.

Warriorrrr reviewed (commented):
I think a new rotation wrapper class would be good for this, but will want see what others think first.

NonSwag committed:
Merge branch 'PaperMC:main' into rotation-argument

Commit:
Add Rotation abstraction and implementation

Introduce the `Rotation` interface and its `RotationImpl` implementation to represent orientation with yaw and pitch. This provides a foundation for handling rotational data across the API.

Commit:
Add rotation management methods to Location class

Introduce `addRotation`, `subtractRotation`, and `setRotation` methods to allow manipulation of `Rotation` objects directly within the `Location` class. These methods enhance code clarity and reduce repetitive manual handling of yaw and pitch adjustments.

Commit:
Update rotation handling to use `Rotation` instead of `Vector2f`

Replaced `Vector2f` with the new `Rotation` API in relevant methods and interfaces. This improves type clarity and aligns with the updated math utility framework. Changes include method signatures and implementations for better consistency.

Commit:
Add JavaDoc comments to Rotation interface

Added JavaDoc descriptions for the `pitch()` and `yaw()` methods, as well as the interface itself. This improves code readability and provides clear documentation for developers.

Commit:
Add `@NonNull` annotations to Rotation parameters in Location

Commit:
use jetbrains annotation for consistency

Warriorrrr commented on paper-api/src/main/java/io/papermc/paper/math/Rotation.java:
```diff
@@ -0,0 +1,23 @@
+ * Represents a rotation with specified pitch and yaw values.
+ */
+@NullMarked
+public interface Rotation {
```
There should be some way to obtain an instance of this if we want it to be useful outside of commands, RotationImpl can be moved into the math package in api and a static method can be added in this class, the same is currently done in Position for example.

Warriorrrr commented on paper-server/src/main/java/io/papermc/paper/util/RotationImpl.java:
```diff
@@ -0,0 +1,7 @@
+package io.papermc.paper.util;
+
+import io.papermc.paper.math.Rotation;
+import org.bukkit.util.Vector;
```
leftover unused import here

NonSwag committed:
Merge branch 'PaperMC:main' into rotation-argument

Commit:
Add factory method for creating Rotation instances

Commit:
Move RotationImpl to the paper-api module

Commit:
make location implement rotation

Commit:
Revert "make location implement rotation"

This reverts commit 52dbef2210d2ffb6c789a6e305d32a3ad030c1f1.

Warriorrrr commented on paper-server/src/main/java/io/papermc/paper/util/MCUtil.java:
```diff
@@ -23,6 +23,7 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.phys.Vec2;
```
Unused import

Warriorrrr commented on paper-api/src/main/java/io/papermc/paper/command/brigadier/argument/resolvers/RotationResolver.java:
```diff
@@ -0,0 +1,17 @@
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import io.papermc.paper.math.Rotation;
+import org.jetbrains.annotations.ApiStatus;
+import org.joml.Vector2f;
```
Unused import

Warriorrrr reviewed (approved):
Has some leftover imports but rest looks good

Commit:
Remove unused imports
